core-bundle/src/Resources/contao/classes/FrontendTemplate.php

```diff
@@ -348,40 +348,62 @@ private function setCacheHeaders(Response $response)
 		/** @var PageModel $objPage */
 		global $objPage;
 
+		// Do not cache the response if caching was not configured at all or disabled explicitly
 		if (($objPage->cache === false || $objPage->cache < 1) && ($objPage->clientCache === false || $objPage->clientCache < 1))
 		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
+			$response->headers->set('Cache-Control', 'no-cache, no-store');
 
-			return $response->setPrivate();
-		}
-
-		// Do not cache the response if a user is logged in or the page is protected
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
-		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
-
-			return $response->setPrivate();
+			return $response->setPrivate(); // Make sure the response is private
 		}
 
+		// Private cache
 		if ($objPage->clientCache > 0)
 		{
 			$response->setMaxAge($objPage->clientCache);
+			$response->setPrivate(); // Make sure the response is private
 		}
 
+		// Shared cache
 		if ($objPage->cache > 0)
 		{
-			$response->setSharedMaxAge($objPage->cache);
-		}
+			$response->setSharedMaxAge($objPage->cache); // Automatically sets the response to public
+
+			// We do vary on cookie for the responses that are cacheable for
+			// the shared cache, which will cause reverse proxies to never load
+			// a response from the cache if the _request_ contains a cookie on
+			// the next request.
+			//
+			// NOTE: We do not generate a cache entry for every response
+			// containing a cookie! Responses containing a cookie will always
+			// be private (see earlier in this method).
+			//
+			// However, we want to be able to force the reverse proxy to still
+			// load the response from the cache even if there is a cookie on
+			// the request in case it was configured so by the user. A typical
+			// example for this would be if you have a member section in a
+			// completely different page tree than your "public" website.
+			//
+			// You can configure the public website to force the cache, which
+			// will ensure that logged in members still get the cached entries
+			// served when they visit it. However, you have to be very careful
+			// in this case, e.g. you cannot share any front end module that
+			// shows different content depending on whether you are logged in
+			// or not in both trees (such as a navigation or login module).
+			//
+			// Exception to this rule is if those modules support ESI requests,
+			// in which case they can handle their own caching headers and only
+			// set "Cache-Control: public" if no member is logged in.
```

Comment:
This explanation is kind of longish and not necessarily on the spot (no hard feelings). Maybe we can rework it?

Reply:
Proposals? It is complex so it should be long and detailed :)

Reply:
```
CURRENT:
    // We do vary on cookie for the responses that are cacheable for
    // the shared cache, which will cause reverse proxies to never load
    // a response from the cache if the _request_ contains a cookie on
    // the next request.

PROPOSED:
    // We vary on cookies if a response is cacheable by
    // the shared cache, so a reverse proxy does not load
    // a response from the cache if the _request_ contains a cookie.


CURRENT:
    // NOTE: We do not generate a cache entry for every response
    // containing a cookie! Responses containing a cookie will always
    // be private (see earlier in this method).

PROPOSED:
    // This DOES NOT mean that we generate a cache entry for every response
    // containing a cookie! Responses containing cookies will always
    // be private (see earlier in this method).


CURRENT:
    // However, we want to be able to force the reverse proxy to still
    // load the response from the cache even if there is a cookie on
    // the request in case it was configured so by the user. A typical
    // example for this would be if you have a member section in a
    // completely different page tree than your "public" website.
    //
    // You can configure the public website to force the cache, which
    // will ensure that logged in members still get the cached entries
    // served when they visit it. However, you have to be very careful
    // in this case, e.g. you cannot share any front end module that
    // shows different content depending on whether you are logged in
    // or not in both trees (such as a navigation or login module).
    //
    // Exception to this rule is if those modules support ESI requests,
    // in which case they can handle their own caching headers and only
    // set "Cache-Control: public" if no member is logged in.

PROPOSED:
    // However, we want to be able to force the reverse proxy to still
    // load a response from cache – even if there is a cookie on
    // the request – in case the admin has configured it so. A typical
    // example for this would be a protected member section in a
    // completely different page tree than the public website.
    // You can configure the public website to force loading from cache, which
    // ensures that logged in members still get served the cached entries.
    //

(The next paragraph is the most confusing one for me.)
    // However, you have to be very careful
    // in this case! You e.g. cannot share a front end module that
    // shows different content depending on whether you are logged in
    // or not in both trees (such as a navigation or login module).
    //
    // An exception to this rule is if those modules support ESI requests,
    // in which case they can handle their own caching headers and only
    // set "Cache-Control: public" if no member is logged in.
```

Reply:
I've updated and shortened the comment. Wdyt?